drivers/flash/flash_sam0.c

```diff
@@ -163,8 +173,8 @@ static int flash_sam0_write_page(const struct device *dev, off_t offset,
 	flash_sam0_wait_ready();
 
 	/* Ensure writes happen 32 bits at a time. */
-	for (; src != end; src++, dst++) {
-		*dst = UNALIGNED_GET((uint32_t *)src);
+	for (; len > 0; len -= sizeof(*src)) {
+		*dst++ = UNALIGNED_GET((uint32_t *)src++);
```

Comment:
`UNALIGNED_GET` is a macro, don't do `++` with a macro parameter.

Reply:
@de-nordic, very nice catch !

Reply:
@de-nordic, what was your thought with end?  It is possible as with the byte implementation, as @attie-argentum mentioned previously, for the length to not be intervals of sizeof(uint32_t)?

Reply:
If `len` is not a multiple of 4, then you'll already be losing the last portion of the write... in fact, you cannot write 3 bytes (only 1, 2 or 4 bytes IIRC). This is guarded against by A) the comment on `flash_sam0_write_page()`, and B) the use of `ROW_SIZE` and `(len % FLASH_WRITE_BLK_SZ) != 0` (depending on the route in).

The issue here is that [`UNALIGNED_GET()` is a macro](https://github.com/zephyrproject-rtos/zephyr/blob/9ca543d7ae2a8bed8b4ecfd3379b3e222b81b8ae/include/zephyr/toolchain/gcc.h#L125-L132), and thus may evaluate the `src++` more than once(!)

I'd prefer to keep the original, with a patch on the `end` variable:

```c
const uint32_t *src = data;
const uint32_t *end = src + (len / sizeof(*src));

// ...

for (; src != end; src++, dst++) {
    *dst = UNALIGNED_GET((uint32_t *)src);
}
```

Reply:
@attie-argentum, I am following now.  Thanks for the explanation.

Reply:
Great! no worries 😊